graphing/src/cloudevolution/views.py

```diff
@@ -139,12 +190,51 @@ def dilutions(request, experiment):
 		# All vials were chemostats or not used
 		efficiency = None
 
+	# Calculate consumption of last bottle
+	bottles = []
+	bottle_info = []  # Stores info displayed in "See bottle setup"
+	bottle_data = open(bottle_file).readlines()[1:]
+
+	# Get experiment start time
+	with open(expt_pickle, 'rb') as f:
+		expt_start = pickle.load(f)[0]
+
+	if not bottle_data:
+		bottle_info = None
+	else:
+		# bottleID	vials	volume0	timestamp0 ... volumeN	timestampN
+		for c, bot in enumerate(bottle_data):
+			# Extract data and store as lists for Django
+			bot = bot.strip("\n").split("\t")
+			bottle_info.append(bot[:2] + bot[-2:])   # ID	vials	volumeN	timestampN
+			# Sum media consumption of vials connected to each bottle
+			media = 0
+			for v in bot[1].split(","):
+				if v:
+					# get bottle timestamp
+					tstamp = time.mktime(time.strptime(bot[-1], "%d/%m/%Y %H:%M"))  # Timestamp in seconds since epoch
+					tstamp -= expt_start  # Timestamp in seconds since start of experiment
+					# get data slice from timestamp to present
+					pump_dir = os.path.join(evolver_dir, expt_subdir[0], experiment, "pump_log", f"vial{v}_pump_log.txt")
+					data = np.genfromtxt(pump_dir, delimiter=',', skip_header=2)
+					try:
+						bottle_consumption = sum(sum(data[np.where(data[:, 0] > tstamp), 1])) * cal[0, int(v)] / 1000
+					except IndexError:
+						bottle_consumption = -1  # Error when slicing the data
+				else:
+					bottle_consumption = -2  # Bottle has no vials assigned
```

Comment:
Are these negative flags used elsewhere? You could just make custom exceptions. Both of these cases seem like there is either something wrong with the data or something wrong with how the user configured the file. In both cases the user should be informed (some sort of indicator on the page).